### PR TITLE
Enabled edition of translations messages

### PR DESCRIPTION
**Pull request by thierrymarianne**
| Questions | Answers |
| --- | --- |
| Branch? | develop |
| Description? | List translations messages and enable edition in the back-office (International > Translations) |
| Type? | bug fix |
| Category? | BO |
| BC breaks? | no |
| Deprecations? | no |
| Fixed ticket? |  |
| How to test? | Edit a translation message and check its corresponding usages in the application |

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Admin/TranslationsController.php [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+     * @Template
+     * @param Request $request
+     * @return array Template vars
+     */
+    public function listAction(Request $request)
+    {
+        $catalogue = $this->getTranslationsCatalogue($request);
+        $translationsTree = $this->makeTranslationsTree($catalogue);
+
+        return ['translationsTree' => $translationsTree];
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/TranslationsController.php]: we used to keep the old school array() notation, sorry :(

File: src/PrestaShopBundle/Controller/Admin/TranslationsController.php [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            throw new \Exception('The legacy to standard locales JSON could not be decoded', $jsonLastErrorCode);
+        }
+
+        return $legacyToStandardLocales;
+    }
+
+    /**
+     * @return string
+     */
+    protected function getResourcesDirectory()
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/TranslationsController.php]: nice helper, what do you think about moving it directly to ``FrameworkBundleAdminController`` class ?

File: src/PrestaShopBundle/Resources/config/admin/routing_international.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+admin_international_translations_list:
+    path: /translations/list
+    methods:  [POST]
+    defaults:
+        _controller: PrestaShopBundle:Admin/Translations:list
+        _legacy_controller: AdminTranslations
+        theme: classic
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/config/admin/routing_international.yml]: I'm wondering.. why do you need both ``theme`` and ``type`` defaults properties, did I miss something ?

File: src/PrestaShopBundle/Twig/TranslationsExtension.php [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    public function makeSubtree($tree, $level = 3)
+    {
+        $output = '';
+        $messagesSubtree = $this->hasCamelizedDomain($tree);
+
+        if ($messagesSubtree) {
+            $camelizedDomain = $tree['__camelized_domain'];
+            unset($tree['__camelized_domain']);
+
+$inputTemplate = <<<FORM
```
Comment (Shudrum) [inline src/PrestaShopBundle/Twig/TranslationsExtension.php]: I really don't like `HEREDOC`s, is it possible to get this template another way?

**Comment by Shudrum** [comment]
:+1:

File: src/PrestaShopBundle/Controller/Admin/TranslationsController.php (hunk 1 above)
Comment (thierrymarianne) [inline src/PrestaShopBundle/Controller/Admin/TranslationsController.php]: Understood, no problem

File: src/PrestaShopBundle/Resources/config/admin/routing_international.yml (hunk 3 above)
Comment (thierrymarianne) [inline src/PrestaShopBundle/Resources/config/admin/routing_international.yml]: Indeed, they are not required anymore. I will remove them as well from the edit section of the BO `Translations` page.  Thanks

File: src/PrestaShopBundle/Twig/TranslationsExtension.php (hunk 4 above)
Comment (thierrymarianne) [inline src/PrestaShopBundle/Twig/TranslationsExtension.php]: I would have rendered a twig template before inclusion but being in an Twig extension, there would be a circular dependency exception. I will revert to using variable concatenation though. Thanks.

Comment (Shudrum) [inline src/PrestaShopBundle/Twig/TranslationsExtension.php]: Thank you!

**Comment by mickaelandrieu** [comment]
Comments have been answered => code reviewed ;)

File: src/PrestaShopBundle/Twig/TranslationsExtension.php (hunk 4 above)
Comment (thierrymarianne) [inline src/PrestaShopBundle/Twig/TranslationsExtension.php]: I've moved the snippets with https://github.com/PrestaShop/PrestaShop/pull/5982/commits/67444e2f6202cab6ca3088796f6625dc848362f0 to views even though Twig is not relied upon to render them yet

**Comment by vincentbz** [comment]
Tested and ok, thanks !
The translation is changed, the cache is cleared so that the modification is visible immediately.

**Comment by mickaelandrieu** [comment]
@thierrymarianne first pull request merged ! :tada:

**Comment by thierrymarianne** [comment]
@mickaelandrieu @vincentbz @Shudrum Thanks!

**Comment by AlexEven** [comment]
Thank you @thierrymarianne !

I had a look and I have an issue with a few strings when I reset it (test in French).

**Admin > Dashboard > Notifications**
Key: This is different from the shop domain name set in the Multistore settings: "%s".
Translation: "Ce nom est différent du nom de domaine configuré dans les paramètres "Multiboutique" : "%s"."

I changed "Multiboutique" to "tralala" (why not?)
Then clicked on "Reset", and I lost half the sentence, I know only have "Ce nom est différent du nom de domaine configuré dans les paramètres ".

**Parameters > Notifications**
Key: "The "Backups" directory located in the admin directory must be writable (CHMOD 755 / 777)."
When I reset after modification I have "Le dossier " instead of "Le dossier "Backups" dans le répertoire "admin" doit être accessible en écriture (CHMOD 755 / 777)."

I checked other strings which are ok, is this to do with the ""?

Thanks!

**Comment by Shudrum** [comment]
I think so!

Thank you @AlexEven!

**Comment by thierrymarianne** [comment]
@Shudrum @AlexEven I believe translation values column in the database should be of `text` type instead of `varchar(255)`.

**Comment by AlexEven** [comment]
Also I'm missing some folders it seems. Not sure if it's linked to my install or what, but could you please check?

**Admin > Modules**,  we miss "Help" and "Notifications"
image

**Admin > Navigation**, we miss "Menu"
image

**Admin > Notification**, we miss "Info"

**Orders > Customers** should be one entity, domain is "OrdersCustomers" not Orders.Customers (I know it's not perfect^^)

**Admin > Stat** is missing

**Modules > Bank > Wire** should be BankWire
Same goes with Block/Adverstising, Best/Sellers, CategoryTree, etc.:
image

**Shop >Theme** we're missing all the subsections: Actions, Catalog, Checkout, CustomerAccount.

@thierrymarianne We can talk about tomorrow if you want, and show you what we want to do with the new translation system.

Thanks!

**Comment by thierrymarianne** [comment]
@AlexEven I've opened another pull-request (https://github.com/PrestaShop/PrestaShop/pull/6026) tackling most of the issues above.

With regards to `Admin > Modules`, "Help" and "Notifications" subdomains are missing as the corresponding translation dictionaries don't contain any messages in the language pack I've received as reference.

Same goes for `Admin > Navigation > Menu`

The dictionary itself is missing for `Admin > Stat`

The domains containing multiple words above are not broken anymore into multiple subdomains but it currently requires the maintenance of a list for such domains with multiple words.

**Comment by Shudrum** [comment]
Ping @julienbourdeau about the missing domains!

**Comment by thierrymarianne** [comment]
@Shudrum after checking this morning with @julienbourdeau, its appeared crowdin doesn't return translations key for missing translations. We're going to reach their support team in order to have a fix of the build feature.

**Pull request merged**